File: 0203-remove-linked-list-elements/0203-remove-linked-list-elements.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeElements(ListNode* head, int val) {
        if (head == nullptr)
            return nullptr;
        while (head != nullptr && head->val == val) {
            ListNode* todel = head;
            head = head->next;
            delete todel;
        }
        ListNode* temp = head;
        if(temp == nullptr) return NULL;
        while (temp->next != nullptr) {
            if (temp->next->val == val) {
                ListNode* todel = temp->next;
                temp->next = temp->next->next;
                delete todel;
            } else {
                temp = temp->next;
            }
        }
        return head;
    }
};

File: 0203-remove-linked-list-elements/0203-remove-linked-list-elements.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 * int val;
 * ListNode next;
 * ListNode() {}
 * ListNode(int val) { this.val = val; }
 * ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

// 1 1 1 2 3 5 1 null
// target = 1
//
class Solution {
    public ListNode removeElements(ListNode head, int val) {
        if (head == null)
            return null;
        ListNode temp = head;

        while (temp != null && temp.val == val)
            temp = temp.next;

        head = temp;
        if(head == null)    return null;

        while (temp.next != null) {
            if (temp.next.val == val) {
                temp.next = temp.next.next;
            } else {
                temp = temp.next;
            }
        }

        return head;
    }
}